dmoj/executors/gcc_executor.py

```diff
@@ -27,23 +27,27 @@ class GCCExecutor(CompiledExecutor):
     arch = 'gcc_target_arch'
     has_color = False
 
-    def create_files(self, problem_id, main_source, **kwargs):
-        aux_sources = kwargs.get('aux_sources', {})
-        fds = kwargs.get('fds', None)
-        writable = kwargs.get('writable', (1, 2))
-        aux_sources[problem_id + self.ext] = main_source
+    def __init__(self, problem_id, main_source, **kwargs):
+        super(GCCExecutor, self).__init__(problem_id, main_source, **kwargs)
+
+        self.source_dict = kwargs.get('aux_sources', {})
+        self.source_dict[problem_id + self.ext] = main_source
+        self.defines = kwargs.get('defines', [])
 
-        sources = []
-        for name, source in aux_sources.items():
+    def create_files(self, problem_id, main_source, **kwargs):
+        self.source_paths = []
+        for name, source in self.source_dict.items():
             if '.' not in name:
                 name += self.ext
             with open(self._file(name), 'wb') as fo:
                 fo.write(utf8bytes(source))
-            sources.append(name)
-        self.sources = sources
-        self._fds = fds
-        self._writable = writable
-        self.defines = kwargs.get('defines', [])
+            self.source_paths.append(name)
+
+    def get_binary_cache_key(self):
+        key_components = ([self.problem, self.get_command(), self.get_march_flag()] +
+                          self.get_defines() + self.get_flags() + self.get_ldflags() +
+                          list(self.source_dict.values()))
+        return ''.join(key_components)
```

Comment:
Is it... wise... to have it be so long? Maybe consider SHA384?

Reply:
Also class name? I feel like your current solution could return the wrong class, if say, the flags are controlled by... environment.